chat message history: introduce interface type, add new history package

mvrilo opened the pull request:
- Introducing a `ChatMessageHistory` interface:
```go
type ChatMessageHistory interface {
    Messages() ([]ChatMessage, error)
    AddMessage(ChatMessage) error
    Clear() error
}
```

- Renamed `ChatMessageHistory` to `SimpleChatMessageHistory` as in-memory implementation
- Added one more example llm chain example using buffer and history

References:
https://github.com/hwchase17/langchainjs/tree/main/langchain/src/stores/message
https://github.com/hwchase17/langchain/tree/master/langchain/memory/chat_message_histories

### PR Checklist

- [x] Read the [Contributing documentation](https://github.com/tmc/langchaingo/blob/main/CONTRIBUTING.md).
- [x] Read the [Code of conduct documentation](https://github.com/tmc/langchaingo/blob/main/CODE_OF_CONDUCT.md).
- [x] Name your Pull Request title clearly, concisely, and prefixed with the name of primarily affected package you changed according to [Good commit messages](https://go.dev/doc/contribute#commit_messages) (such as `memory: add interfaces for X, Y` or `util: add whizzbang helpers`).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Describes source of new concepts.
- [x] References existing implementations as appropriate.
- [x] Contains test coverage for new functions.
- [x] Passes all [`golangci-lint`](https://golangci-lint.run/) checks.

alvarowolfx commented on memory/history/simple.go:
```diff
@@ -0,0 +1,55 @@
+var _ schema.ChatMessageHistory = (*SimpleChatMessageHistory)(nil)
+
+// SimpleChatMessageHistory is a struct that stores chat messages.
+type SimpleChatMessageHistory struct {
```
the `SimpleChatMessageHistory` is the equivalent to the original In-memory implementation. Should we just call it InMemoryChatMessageHistory instead ?

mvrilo replied:
I kept the "simple" prefix because the in-memory `Memory` implementation is also referenced as "simple" here:
https://github.com/tmc/langchaingo/blob/main/memory/simple.go#L9

alvarowolfx reviewed (commented):
Added a comment about naming the In Memory implementation. Can we add this history memory implementation to one of the chat samples/examples ?

mvrilo commented:
By default the [`Buffer` memory impl](https://github.com/mvrilo/langchaingo/blob/history-interface/memory/buffer.go#L32) already uses the `SimpleChatMessageHistory`.

Here's a quick usage, setting it manually:

```
mem := memory.NewBuffer()
mem.ChatHistory = history.NewSimpleChatMessageHistory()

llmChain := chains.NewLLMChain(llm, prompts.NewPromptTemplate("How old is the planet {{.planet}}?", []string{"planet"}))
llmChain.Memory = mem

ctx := context.Background()
out, err := chains.Call(ctx, llmChain, map[string]any{"planet": "Earth"})
if err != nil {
    return err
}

fmt.Println(out["text"])
```

I wanted to keep the PR small but I can add this in the examples, no problem.

Another relevant change that I could add to this PR, before adding the example, is changing the `buffer.New()` to `buffer.New(history schema.ChatMessageHistory)` in the `Buffer` impl.

mvrilo commented:
Updated my PR changing interface methods to return errors, added one more example and updated the `Buffer` initializer to `buffer.New(schema.ChatMessageHistory)`.

tmc commented:
Why not just accept the general ChatMessage interface vs the more specific methods here?

mvrilo commented:
Where exactly @tmc? You mean for the `buffer.NewBuffer` to accept a `[]schema.ChatMessage` instead of the new type?

tmc commented:
> Where exactly @tmc? You mean for the `buffer.NewBuffer` to accept a `[]schema.ChatMessage` instead of the new type?

Sorry I missed this message. 

I mean why not something like:

```go
type ChatMessageHistory interface {
    Messages() ([]ChatMessage, error)
    AddMessage(ChatMessage) error
    Clear() error // note: I'm not sure clear is necessary here.
}
```

mvrilo commented:
Thanks! The other implementations had the AddMessage for each role, that's why I kept that way. I'll update the PR.

tobiade commented:
Actually, I just realised I had a brain-fart moment and could have just done:
```go
rc := textsplitter.NewRecursiveCharacter()
rc.ChunkSize = 1024
rc.ChunkOverlap = 0
```

I'm happy to close this if you think functional options isn't that important here.
